File: DMMPVP/core/src/main/java/com/zenyte/game/content/skills/construction/objects/portalchamber/Portal.java
package com.zenyte.game.content.skills.construction.objects.portalchamber;

import com.zenyte.game.content.achievementdiary.diaries.MorytaniaDiary;
import com.zenyte.game.content.skills.construction.Construction;
import com.zenyte.game.content.skills.construction.ObjectInteraction;
import com.zenyte.game.content.skills.construction.RoomReference;
import com.zenyte.game.content.skills.construction.TeleportPortal;
import com.zenyte.game.world.entity.player.Player;
import com.zenyte.game.world.object.WorldObject;

import java.util.ArrayList;

/**
 * @author Kris | 25. veebr 2018 : 22:58.14
 * @see <a href="https://www.rune-server.ee/members/kris/">Rune-Server profile</a>}
 * @see <a href="https://rune-status.net/members/kris.354/">Rune-Status profile</a>}
 * TODO: Restrictions for leaving through portal.
 */
public final class Portal implements ObjectInteraction {
	@Override
	public Object[] getObjects() {
		final ArrayList<Object> list = new ArrayList<Object>(TeleportPortal.VALUES.length * 3);
		for (final TeleportPortal portal : TeleportPortal.VALUES) {
			for (int i : portal.getPortalIds()) list.add(i);
		}
		return list.toArray(new Object[list.size()]);
	}

	@Override
	public void handleObjectAction(Player player, Construction construction, RoomReference reference, WorldObject object, int optionId, String option) {
		if (option.equals("enter")) {
			final TeleportPortal portal = TeleportPortal.MAP.get(object.getId());
			if (portal == null) {
				return;
			}
			if (portal.equals(TeleportPortal.KHARYRLL_PORTAL)) {
				player.getAchievementDiaries().update(MorytaniaDiary.ENTER_KHARYRLL_PORTAL);
			}
			player.setLocation(portal.getTeleport().getDestination());
		}
	}
}

File: DMMPVP/plugins/excluded/theatreofblood/src/main/kotlin/com/zenyte/game/content/theatreofblood/room/sotetseg/object/Portal.kt
package com.zenyte.game.content.theatreofblood.room.sotetseg.`object`

import com.zenyte.game.content.theatreofblood.room.sotetseg.ShadowRealmRoom
import com.zenyte.game.content.theatreofblood.room.sotetseg.npc.Sotetseg
import com.zenyte.game.world.entity.player.Player
import com.zenyte.game.world.`object`.ObjectAction
import com.zenyte.game.world.`object`.ObjectId
import com.zenyte.game.world.`object`.WorldObject
import com.zenyte.game.world.region.RegionArea

/**
 * @author Tommeh
 * @author Jire
 */
internal class Portal : ObjectAction {

    override fun handleObjectAction(
        player: Player,
        `object`: WorldObject,
        name: String,
        optionId: Int,
        option: String
    ) {
        if (!option.equals("Enter", ignoreCase = true)) return
        val area: RegionArea = player.area as? ShadowRealmRoom ?: return
        val realm = area as ShadowRealmRoom
        if (realm.completed) return
        val sotetseg = realm.boss as Sotetseg
        realm.completed = true
        sotetseg.completeMaze(player)
    }

    override fun getObjects() = Portal.objects

    companion object {

        private val objects = arrayOf(ObjectId.PORTAL_33037)

    }

}